src/Interpreters/HashJoin.cpp

```diff
@@ -879,6 +882,28 @@ bool HashJoin::addBlockToJoin(const Block & source_block_, bool check_limits)
         }
     }
 
+    auto max_total_bytes = table_join->sizeLimits().max_bytes;
```

Comment:
But it will work only if the limit for JOIN size is defined. Let's also take max_memory_usage into account. Also, let's use the current memory usage for the query.

Reply:
For comparison, see how we do it for external GROUP BY.

Reply:
How to combine this two metrics?
We can compare memory tracker delta with `max_memory_usage` and `current_bytes_in_join` with `max_bytes_in_join` separately (as done in recently pushed commit).
Or compare `max(memory_tracker_delta, current_bytes_in_join)` and `min(max_bytes_in_join, max_bytes_in_join)`.